### PR TITLE
Change `extends` in `overrides` to be consistent with `plugins` behaviour

### PR DESCRIPTION
**Pull request by jasikpark**
- Adds docs describing how different properties get merged in `overrides`
- Merges `extends` arrays together in `overrides`
- Adds a test for that `extends` merging functionality

### CODE REVIEW

**Comment by jeddy3** [comment]
@jasikpark Thanks for opening this pull request to show the current behaviour.

> If this is the intended behavior

I've been looking back through some issues (https://github.com/stylelint/stylelint/pull/5603, https://github.com/stylelint/stylelint/issues/5656 etc) and it doesn't look like we ever discussed what the intended behaviour should be. 

It feels safe to merge `plugins` as it simply increases the pool of available rules. `extends`, on the other hand, can change what rules are turned on and off. However, having consistent behaviour for both may make the most sense.

Although the current behaviour is undocumented, we can't know how many users rely on it.

Anyone else got thoughts on this?

Whatever we decide, let's update [our docs](https://stylelint.io/user-guide/configure#overrides) to describe the behaviour for each of the most used configuration properties: `rules`, `customSyntax`, `extends` and `plugins`.

At the moment, I believe it's:

- `rules` - merge
- `customSyntax` - replace
- `extends` - replace
- `plugins` - merge

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for summarizing the issues.

> `extends`, on the other hand, can change what rules are turned on and off. However, having consistent behaviour for both may make the most sense.

I agree with the idea that `extends` should cover merging like `plugins`.
And, also updating our docs. 👍🏼

**Comment by jeddy3** [comment]
OK, let's make the `extends` behaviour consistent with `plugins`. It was an oversight, so shall we classify as a bug fix?

@jasikpark Feel free to fix the behaviour in this pull request and update the docs. 

For the docs, we can add a bullet at the [end of this section](https://stylelint.io/user-guide/configure#overrides).

- Configuration properties with a single value, e.g. `customSyntax`, will be overridden whereas everything else, e.g. `extends`, `plugins`, `rules` etc., will be merged.

**Comment by jasikpark** [comment]
Okay, pushed some basics docs & merging `extends`. What is the explicit list of things that should be merged vs replaced?

**Comment by jasikpark** [comment]
looks like `rules` are not deduplicated, though they are merged

https://github.com/jasikpark/stylelint/blob/7ac953e87cef7da2b7264220fd766b361e4ee5b5/lib/augmentConfig.js#L236-L246

**Comment by jasikpark** [comment]
I guess everything in 

https://github.com/stylelint/stylelint/blob/cf04b2be5b7e14ec3f9c633a9b1a7d1653faa920/types/stylelint/index.d.ts#L36-L57

that is an array should be merged right? since technically any of those can be passed to overrides?

**Comment by jasikpark** [comment]
might be worth looking to eslint's behavior here as well https://eslint.org/docs/latest/user-guide/configuring/configuration-files#how-do-overrides-work

**Comment by jasikpark** [comment]
Seems that `extends`, `plugins`, `processors`, `processorFunctions`, `ignoreFiles`, `codeProcessors`, `resultProcessors` are arrays and might be merged similarly to the existing `plugins` behavior

`rules` and `pluginFunctions` are objects containing a variable number of properties, which could also be merged potentially

`ignorePatterns`, `quiet`, `defaultSeverity`, `ignoreDisables`, `report*`, and `customSyntax` are javscript built-in datatypes like boolean and string and should be replaced.

I'd suggest maybe this PR can just alter `extends` behavior and then the other config values can be updated in a second PR?

**Comment by jasikpark** [comment]
Since I'm not sure what the semantics are of all these config values

**Comment by jeddy3** [comment]
@jasikpark Thanks for making the changes and for digging more into the overrideable options.

> I'd suggest maybe this PR can just alter extends behavior and then the other config values can be updated in a second PR?

That sounds best. The general behaviour you outline in https://github.com/stylelint/stylelint/pull/6380#issuecomment-1267722440, SGTM. 

We're going to remove Processors in the next major release. That leaves the following as user-facing options:

- `extends`
- `plugins`
- `ignoreFiles`
- `quiet`
- `defaultSeverity`
- `ignoreDisables`
- `report*`
- `customSyntax`

If we haven't already, it'd be good to have tests for these.

I've added a changeset entry. I think this pull request is ready for review.

File: docs/user-guide/configure.md
Comment (jeddy3) [inline docs/user-guide/configure.md:301]:
```suggestion
- `customSyntax` will be replaced by overrides.
- `plugins`, `extends`, `rules`, etc. will be appended.
```

As we now have tests for these, let's go with them for now.

**Review by jeddy3 — CHANGES_REQUESTED** [review]
> Maybe I should add tests + docs that describe where we are today && then another PR can fix everything?

Would you mind doing that in another pull request?

I've made a suggestion to tweak the docs for this pull request.

File: .changeset/grumpy-scissors-watch.md
Comment (jasikpark) [inline .changeset/grumpy-scissors-watch.md:2]: is this a patch when it's almost certainly a breaking change for existing configs? ig we're saying it's a bugfix?

Comment (ybiquitous) [reply .changeset/grumpy-scissors-watch.md:2]: The PR change might be a potential breaking change for some people, but I believe we'd like to avoid as frequent major releases as possible. Because we plan to have various features and breaking changes on the next major version, so the merge of this PR would be blocked until the next major version would be prepared.

For people who would like to keep the current behavior, it might be more kind if there would be a code example like this:

```md
`overrides.extends` is fixed to make consistency with the `overrides.plugins` behavior and be merged not replaced.
If you would like to keep the previous behavior, you would need to change your config as below:
```

```diff
{
- "extends": ["config-a"]
  "overrides": [
    {
      "rules": ["*.module.css"],
      "extends": ["config-b"]
    },
+   {
+     "rules": ["*.css"],
+     "extends": ["config-a"]
+   },
  ]
}
```

**Review by ybiquitous — APPROVED** [review]
Thanks, LGTM 👍🏼

**Review by jeddy3 — APPROVED** [review]
LGTM, thank you!

We can merge once we start actively working on a `v15` branch. 

I've created a tracking issue for preparing 15.0.0 (https://github.com/stylelint/stylelint/issues/6409).

**Comment by ybiquitous** [comment]
I've updated the PR base branch to `v15`. Can we merge this PR? @jeddy3

**Comment by jeddy3** [comment]
I've moved the migrating information out of the changeset and into a migration guide doc.

> I've updated the PR base branch to v15. Can we merge this PR? @jeddy3

Yes.

**Comment by jeddy3** [comment]
This merge behaviour is aligned with how a flat config would work, and so is a step in the right direction for us.

[From this article](https://eslint.org/blog/2022/08/new-config-system-part-2/#goodbye-extends%2C-hello-flat-cascade):

> Inside of the array, ESLint finds all config objects that match the file being linted and merges them together in much the same way that eslintrc did. The only real difference is the merge happens from the top of the array down to the bottom instead of using files in a directory structure.

File: docs/migration-guide/to-15.md
Comment (ybiquitous) [inline docs/migration-guide/to-15.md:3]: Typo?

```suggestion
## Change of `overrides.extends` behavior
```

File: docs/migration-guide/to-15.md
Comment (ybiquitous) [inline docs/migration-guide/to-15.md:20]: Looks like a typo because JSON disallows a trailing comma.

```suggestion
+   }
```

**Review by ybiquitous — APPROVED** [review]
@jeddy3 Thanks so much for writing up the migration guide. There may be a few typos left, but LGTM! 👍🏼

**Comment by jasikpark** [comment]
thx for getting this merge ready 💜

File: docs/migration-guide/to-15.md
Comment (ybiquitous) [inline docs/migration-guide/to-15.md:11]:
```suggestion
- "extends": ["config-a"],
```

**Comment by ybiquitous** [comment]
We can merge this when CI passes. 💪🏼

**Pull request merged**
